src/uproot/source/futures.py

```diff
@@ -469,3 +472,45 @@ def __exit__(self, exception_type, exception_value, traceback):
         self.shutdown()
         self._resource.__exit__(exception_type, exception_value, traceback)
         self._closed = True
+
+
+class LoopExecutor:
+    def __repr__(self):
+        return f"<LoopExecutor at 0x{id(self):012x}>"
+
+    def __init__(self):
+        self._loop = asyncio.new_event_loop()
+        # TODO: remove daemon=True (or not?)
+        self._thread = threading.Thread(target=self._run, daemon=True)
+        self.start()
+
+    def start(self):
+        self._thread.start()
+        return self
+
+    def shutdown(self):
+        self._loop.call_soon_threadsafe(self._loop.stop)
+        self._thread.join()
+
+    def _run(self):
+        asyncio.set_event_loop(self._loop)
+        try:
+            self._loop.run_forever()
+        finally:
+            self._loop.run_until_complete(self._loop.shutdown_asyncgens())
+            self._loop.close()
```

Comment:
We probably need to run cleanups here too, *and* shield these from cancellation

```python
    # Find all running tasks:
    pending = asyncio.all_tasks()

    # Run loop until tasks done:
    loop.run_until_complete(
        asyncio.gather(*[
            asyncio.shield(t) for t in pending
            ]
        )
    )
```

Reply:
This is mostly my fault. I don't understand why there isn't a canned "loop-in-a-thread" in the stdlib or elsewhere.

Reply:
@nsmith I wouldn't ascribe fault — asyncio is a bit of an ugly beast (I use `trio` in my personal projects these days), but it's the one we're stuck with.

The intention with `asyncio` is that you only really have one loop. What we're doing is a bit unconventional, but clearly necessary to handle a blocking high level UX.

Reply:
We effectively want this: https://github.com/python/cpython/blob/3.8/Lib/asyncio/runners.py#L8-L73

Reply:
Ah, yes! The previous error was an unhandled exception, and I would not be surprised if we have another one.